Xtensa asm2 and SMP series, full

andyross opened the pull request:
Rebased, should be ready for merge unless I missed some test failures

Commit:
xtensa: Move register window exception handlers into a separate file

No behavior changes, just code motion.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/timer/xtensa_sys_timer: Correctly declare ISR

The existing hand-written interrupt code is manually calling the timer
ISR, which is sort of silly and about to be replaced.  Correctly
declare the ISR with IRQ_CONNECT() so that a conventional interrupt
handling implementation can find it.  With current code this is a
noop.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Make high priority interrupts optional

Xtensa has a "high priority" class of interrupt levels which ignore
the EXCM bit and can thus interrupt running exception handlers.  These
can't be used for C handlers in the general case[1] because C code
needs to be able to throw window over/underflow exceptions, which are
not reentrant.

But the high priority interrupts might be useful to a carefully
designed application, or to unit tests of low level architecture code.
So make their generation optional with this kconfig option.

[1] ESP-32 has a high priority interrupt for its watchdog, apparently.
    Which is sort of OK given that it never needs to return to the
    interrupted code.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: New asm layer to support SMP

SMP needs a new context switch primitive (to disentangle _swap() from
the scheduler) and new interrupt entry behavior (to be able to take a
global spinlock on behalf of legacy drivers).  The existing code is
very obtuse, and working with it led me down a long path of "this
would be so much better if..."  So this is a new context and entry
framework, intended to replace the code that exists now, at least on
SMP platforms.

New features:

* The new context switch primitive is xtensa_switch(), which takes a
  "new" context handle as an argument instead of getting it from the
  scheduler, returns an "old" context handle through a pointer
  (e.g. to save it to the old thread context), and restores the lock
  state(PS register) exactly as it is at entry instead of taking it as
  an argument.

* The register spill code understands wrap-around register windows and
  can avoid spilling A4-A15 registers when they are unused by the
  interrupted function, saving as much as 48 bytes of stack space on
  the interrupted stacks.

* The "spill register windows" routine is entirely different, using a
  different mechanism, and is MUCH FASTER (to the tune of almost 200
  cycles).  See notes in comments.

* Even better, interrupt entry can be done via a clever "cross stack
  call" I worked up, meaning that the interrupted thread's registers
  do not need to be spilled at all until they are naturally pushed out
  by the interrupt handler or until we return from the interrupt into
  a different thread.  This is a big efficiency win for tiny
  interrupts (e.g. timers), and a big latency win for all interrupts.

* Interrupt entry is 100% symmetric with respect to medium/high
  interrupts, avoiding the problems seen with hooking high priority
  interrupts with the current code (e.g. ESP-32's watchdog driver).

* Much smaller code size.  No cut and paste assembly.  No use of HAL
  calls.

* Assumes "XEA2" interrupt architecture, the register window extension
  (i.e. no CALL0 ABI), and the "high priority interrupts" extension.
  Does not support the legacy processor variants for which we have no
  targets.  The old code has some stuff in there to support this, but
  it seems bitrotten, untestable, and I'm all but certain it doesn't
  work.

Note that this simply adds the primitives to the existing tree in a
form where they can be unit tested.  It does not replace the existing
interrupt/exception handling or _Swap() implementation.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
samples/xtensa-asm2: Unit test for new Xtensa assembly primitives

This sample (which should eventually become a proper test) suite
builds from simple applications of the new primitives to a full
context switch test and interrupt handling suite (based on the
CPU-internal CCOMPARE2 timer).

It's been extraordinarily useful finding regressing as the asm2 code
gets modified and should probably stick around as long as possible.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Fix nano_internal.h inclusion

_Swap() is defined in nano_internal.h.  Everything calls _Swap().
Pretty much nothing that called _Swap() included nano_internal.h,
expecting it to be picked up automatically through other headers (as
it happened, from the kernel arch-specific include file).  A new
_Swap() is going to need some other symbols in the inline definition,
so I needed to break that cycle.  Now nothing sees _Swap() defined
anymore.  Put nano_internal.h everywhere it's needed.

Our kernel includes remain a big awful yucky mess.  This makes things
more correct but no less ugly.  Needs cleanup.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Move some macros

K_NUM_PRIORITIES and K_NUM_PRIO_BITMAPS were defined in
nano_internal.h, but used in only a handful of places.  Move to
kernel_structs.h (somewhat higher up in the hierarchy) to help with
include file cycle-breaking.  Arguably they are a better fit there
anyway.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Add alternative _arch_switch context switch primitive

The existing __swap() mechanism is too high level for some
applications because of its scheduler-awareness.  This introduces a
new _arch_switch() mechanism, which is a simpler primitive that looks
like:

    void _arch_switch(void *handle, void **old_handle_out);

The new thread handle (typically just a stack pointer) is specified
explicitly instead of being picked up from the scheduler by
per-architecture code, and on return the "old" thread handle that got
switched out is returned through the pointer.

The new primitive (currently available only on xtensa) is selected
when CONFIG_USE_SWITCH is "y".  A new C _Swap() implementation based
on this primitive is then added which operates compatibly.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa/asm2: Add a _new_thread implementation for asm2/switch

Implement _new_thread in terms of the asm2 switch mechanism.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Add Kconfig for asm2 layer

The asm2 layer will build alongside the traditional assembly, but the
reverse is not true.  Add a CONFIG_XTENSA_ASM2 to force its use at
runtime and disable the older code.

Note that the older assembly had an initialization function that is
properly part of the timer driver.  Move a C equivalent into the timer
driver itself for now to prevent a build breakage.  Long term we need
to clean that driver up in a bunch of other ways.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Remove arch-specific (and empty) offsets.h

The xtensa arch code had this empty offsets.h header sitting around.
Its name collides with the autogenerated offsets.h, making it
dangerously dependent on include file path order.  Seems to be benign,
but it's freaking me out.  Remove.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Add exception/interrupt vectors in asm2 mode

This adds vectors for all interrupt levels defined by core-isa.h.

Modify the entry code a little bit to select correct linker sections
(levels 1, 6 and 7 get special names for... no particularly good
reason) and to constructed the interrupted PS value correctly (no EPS1
register for exceptions since they had to have interrupted level 0
code and thus differ only in the EXCM bit).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Interrupt generator script and output for qemu & esp32

This python script reads the core-isa.h interrupt definitions (via
running a template file through the toolchain preprocessor to generate
an input file) and emits a fully populated, optimized C handling code
that binary searches only the declared interrupts at a given level and
correctly detects spurious interrupts (and/or incorrect core-isa.h
definitions).

The generated code, alas, turns out not to be any faster than simply
searching the interrupt mask with CLZ (er, NSAU in xtensese), though
it could be faster in theory if the compiler made different choices,
see comments.  But I like this for the robustness of the fully
populated search trees and the checking of level vs. mask.

This simply commits the script output into the source tree, including
some checking code to force a build error if the toolchain changes the
headers incompatibly.  It would be better long term to have these
headers be generated at build time, but that requires more cmake fu
than I have.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Remove _xt_set_exception_handler()

This was a dead API.  Nothing ever used it, it wasn't exposed in any
API headers.  It never appeared in documentation.  It's not
particularly clear why a Zephy app would want to hook
architecture-specific exceptions instead of simply using the portable
error framework anyway. And it's not supported by asm2.  Delete.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Implement _xt_ints_on/off for asm2

Legacy xtensa had a rather complicated implementation of en/disabling
interrupts, owing to the "software priority" feature (which plays
games with INTENABLE and INTLEVEL to allow for interrupts to interrupt
each other outside their normal priorities).  But that's not a Zephyr
feature, it's enabled by a XT_USE_SWPRI value that comes from platform
headers and isn't enabled on any of our boards.  Dead code, basically.

Replace with the obvious implementation when asm2 is in use.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/timer/xtensa_sys_timer: Add init/update hooks for asm2

The earlier xtensa layer put the timer initialization and update
directly into the interrupt handler, which is... weird.  Under asm2,
it's just a regular ISR and needs to do the work in the driver.

Really, this driver needs a bunch of cleanup.  The xtensa CPU timer is
two registers and one ISR: a global cycle count register, and a
compare register that will fire the IRQ when they match.  There is
*way* too much code here.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel, xtensa: Switch-specific thread return value

When using _arch_switch() context switching, the thread return value
is a generic hook and not provided by the architecture.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Fix thread entry point

The stack initilaization was calling the user-provided entry function
directly, which works fine until that function returns, at which point
it will try to unspill A0-A3 from the 16 bytes above the allocated
stack and then "return" to a NULL pointer.

The kernel provides a _thread_entry() function that does cleanup
properly, so use that.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: asm2 needs to honor thread preemption

Forgot to check the thread preemption status when fetching the stack
to restore.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: THREAD_MONITOR hooks for asm2

You'd this feature would be portable, but it's arch-specific.
Initialize the CONFIG_THREAD_MONITOR stuff, placing the __thread_entry
struct (which AFAICT is dead: nothing in the tree actually reads it)
at the top of the stack.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa asm2: Fixup stack alignment at runtime

The API allows any byte count for stack size, and tests in fact check
that a stack with a 499 byte stack works correctly.  No choice, have
to do this at runtime.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Add hook to do register window spills

This macro was already available add an external symbol so C code can
access it (via CALL0 -- it's not and can't be an actual function).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa-asm2: Exception/interrupt handler should check stack sentinel

This got forgotten.  Note that this function is empty if
CONFIG_STACK_SENTINEL is unconfigured.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa-asm2: Handle alloca/movsp exceptions

Xtensa register windows have a special exception that happens when the
stack pointer needs to be moved, but the caller function has already
spilled its registers below it.

I thought these were unexercised in Zephyr code, but they turn out to
be thrown by the existing mem_pool tests when run in the 32-register
qemu environment (but not on 64-register hardwre).  Because the effect
of the exception is to unspill the caller, there is no good way to
handle this in a traditional handler.  Instead put a 5-instruction
stub in front of the user exception handler (i.e. incurring that cost
on every trap and every L1 interrupt) to test before doing the normal
entry.

Works, but would be nicer to optimize this in the future so that only
true alloca exceptions take that cost.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Enable SMP

Now that all the pieces are in place, enable SMP for real:

Initialize the CPU records, launch the CPUs at the end of kernel
initialization, have them wait for a flag to release them into the
scheduler, then enter into the runnable threads via _Swap().

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
esp-32, qemu_xtensa: Use asm2 by default

Set these SoCs to use asm2 by default

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel, esp32: Add _arch_start_cpu API

This is a mostly-internal API to start a secondary system CPU, with an
implementation for the ESP-32 "APP" cpu.  Exposed in kernel.h because
it's plausibly useful for asymmetric MP code managed by an app.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel: Simple test for multiprocessor start API

Starts the second CPU and verifies that it can set a variable while we
spin on the first.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel, esp32: Add SMP kconfig flag and MP_NUM_CPUS variable

Simply define the Kconfig variables in this patch so they can be used
in later patches.  Define MP_NUM_CPUS correctly on esp32.  No code
changes.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Simple spinlock API

Minimal spinlock API based on the existing atomic.h layer.  Usage
works just like irq_lock(), but takes an argument to a specific struct
k_spinlock_t to un/lock.  No attempt at implementing fairness or
backoff semantics.  No attempt made at architecture-specific assembly.

When CONFIG_SMP is not enabled, this code falls back to a zero-size
struct and becomes functionally identical to irq_lock/unlock().

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel: Add spinlock test

Simple test of spinlock semantics.  Bounce between two CPUs locking
and releasing, validating that nothing changes at unexpected times.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Add kswap.h header to unbreak cycles

The xtensa-asm2 work included a patch that added nano_internal.h
includes in lots of places that needed to have _Swap defined, because
it had to break a cycle and this no longer got pulled in from the arch
headers.

Unfortunately those new includes created new and more amusing cycles
elsewhere which led to breakage on other platforms.

Break out the _Swap definition (only) into a separate header and use
that instead.  Cleaner.  Seems not to have any more hidden gotchas.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Move per-cpu _kernel_t fields into separate struct

When in SMP mode, the nested/irq_stack/current fields are specific to
the current CPU and not to the kernel as a whole, so we need an array
of these.  Place them in a _cpu_t struct and implement a
_arch_curr_cpu() function to retrieve the pointer.

When not in SMP mode, the first CPU's fields are defined as a unioned
with the first _cpu_t record.  This permits compatibility with legacy
assembly on other platforms.  Long term, all users, including
uniprocessor architectures, should be updated to use the new scheme.

Fundamentally this is just renaming: the structure layout and runtime
code do not change on any existing platforms and won't until someone
defines a second CPU.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: SMP timer integration

In SMP, the system timer is used for timeslicing on auxiliary CPUs,
but the base system timekeeping via _nano_sys_clock_tick_announce() is
still done on CPU0 only (because the framework isn't prepared for
asynchronous notification yet).  Skip processing on CPU1+.

Also, due to a hardware interaction* that is difficult to work around,
timer initialization on the auxiliary CPUs is done at the very end of
the CPU bringup, just before the swap into the scheduler.  A
smp_timer_init() API has been added for this purpose.

* On ESP-32, enabling the timer seems to result in a near-synchronous
  interrupt being delivered despite my best attempts to keep it
  masked, then blowing things up because the CPU record isn't set up
  to handle it yet.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Create idle threads and interrupt stacks for SMP processors

Simple implementation that caps at 4 CPUs.  Long term we should use
some linker magic to define as many as needed and loop over them
without needlessly increasing data or code size for the tracking.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Make irq_{un}lock() APIs into a global spinlock in SMP mode

In SMP mode, the idea of a single "IRQ lock" goes away.  Long term,
all usage needs to migrate to spinlocks (which become simple IRQ locks
in the uniprocessor case).  For the near term, we can ease the
migration (at the expense of performance) by providing a compatibility
implementation around a single global lock.

Note that one complication is that the older lock was recursive, while
spinlocks will deadlock if you try to lock them twice.  So we
implement a simple "count" semantic to handle multiple locks.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: SMP-aware scheduler

The scheduler needs a few tweaks to work in SMP mode:

1. The "cache" field just doesn't work.  With more than one CPU,
   caching the highest priority thread isn't useful as you may need N
   of them at any given time before another thread is returned to the
   scheduler.  You could recalculate it at every change, but that
   provides no performance benefit.  Remove.

2. The "bitmask" designed to prevent the need to individually check
   priorities is likewise dropped.  This could work, but in fact on
   our only current SMP system and with current K_NUM_PRIOPRITIES
   values it provides no real benefit.

3. The individual threads now have a "current cpu" and "active" flag
   so that the choice of the next thread to run can correctly skip
   threads that are active on other CPUs.

The upshot is that a decent amount of code gets #if'd out, and the new
SMP implementations for _get_highest_ready_prio() and
_get_next_ready_thread() are simpler and smaller, at the expense of
having to drop older optimizations.

Note that scheduler synchronization is unchanged: all scheduler APIs
used to require that an irq_lock() be held, which means that they now
require the global spinlock via the same API.  This should be a very
early candidate for lock granularity attention!

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Report CPU number in exceptions

In SMP contexts it's good to know which CPU blew up.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
esp32: Set CPU pointer on app cpu at startup

It's not enough to wait for a switch, lots of things need this early.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Simplified idle for SMP auxiliary CPUs

A pure timer-based idle won't work well in SMP.  Without an IPI to
wake up idle CPUs out of the scheduler they will sleep far too long
and the main CPU will do all the scheduling of wake-up-and-sleep
processes.  Instead just have the auxilary CPUs do a traditional
busy-wait scheduler in their idle loop.

We will need to revisit an architecture that allows both
wait-for-timer-interrupt idle and SMP.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel: SMP test

Simple SMP test to validate the two threads can be simultaneously
scheduled.  Arranges things such that both threads are at different
priorities and never yield the CPU, so on a uniprocessor build they
cannot be fairly scheduled.  Checks that both are nonetheless making
progress.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/include: Missed nano_internal.h -> kernel_internal.h spots

Update heading naming given recent rename

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

nashif commented:
can you please all other open PRs that are obsolete now?

Commit:
nios2: Add include for _check_stack_sentinel()

This API moved into kswap.h and the resulting warning on this arch got
missed.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa: Fix noreturn attribute on error handlers in asm2

In asm2, the machine exception handler runs in interrupt context (this
is good: it allows us to defer the test against exception type until
after we have done the stack switch and dispatched any true
interrupts), but that means that the user error handler needs to be
invoked and then return through the interrupt exit code.

So the __attribute__(__noreturn__) that it was being decorated with
was incorrect.  And actually fatal, as with gcc xtensa will crash
trying to return from a noreturn call.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
xtensa/asm2: Don't return into dead threads

It's not impossible that something we just handled (e.g. a machine
exception) called k_thread_abort() on our current thread.  Don't try
to return into it, check the DEAD state.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel: Allow k_thread_abort(_current) from ISRs

Traditionally k_thread_abort() of the current thread has done a
synchronous _Swap() to the new context.  Doing this from an ISR has
never worked portably (some architectures can do it, some can't) for
this reason.

But on Xtensa/asm2, exception handlers now run in interrupt context
and it's a very reasonable requirement for them to abort the excepting
thread.

So simply don't swap, but do the rest of the bookeeping, returning to
the calling context.  As a side effect it's now possible to terminate
threads from interrupts, even if they have been interrupted.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
tests/kernel/fatal: Add xtensa/asm2 to the "error returns" family

This test had to special case ARM, where error handlers are not
NORETURN functions.  The xtensa/asm2 layer has the same behavior
(albeit for a different reason).  Add it to the list, and clean up the
explanation a bit.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
misc: Add a FORCE_NO_ASSERT kconfig directive

The mbedtls test is hitting a compiler bug where two subtests will
soft fail on qemu_xtensa when assertions are enabled.  This is despite
the fact that:

+ The failure is entirely internal to the mbedtls suite.
+ The mbedtls code does not use zephyr asserts
+ The mbedtls code does not call into zephyr code that might assert.
+ The behavior persists even when an irq_lock() is held across the
  entire test, ruling out any asserts in interrupt/exception context.
+ And EVEN WHEN the mbedtls library blobs are bytewise identical
  between assert and non-assert cases.

The bug seems to be a layout thing where the mbedtls code behavior
differently based on code address and/or link-time optimizations
(xtensa has a few).

Unfortunately sanitycheck enables assertions by setting CFLAGS
directly and not via kconfig, so we can't fix this by turning the
feature off in an app right now.  This patch adds a simple "override"
flag that can be set by apps like this that hit bugs.

Again, note that zephyr assertions are not used nor needed by this one
test.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/kswap.h: Hook event logger from switch-based _Swap

The new generic _Swap() forgot the event logger hook

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
mpu_stack_guard_test: Disable on xtensa

As with the other non-MPU architectures, we don't want to run this
test.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
esp32: Move hard firmware function addresses to the linker script

This matches the way other firmware-provided functions are done.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

ramakrishnapallala commented on arch/xtensa/core/window_vectors.S:
```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2016 Cadence Design Systems, Inc.
```
2018?

ramakrishnapallala commented on arch/xtensa/core/thread.c:
```diff
@@ -12,6 +12,7 @@
 #include <kernel_structs.h>
 #include <wait_q.h>
 #include <xtensa_config.h>
+#include <nano_internal.h>
```
we changed this name to kernel_internal.h

is it still existing?

ramakrishnapallala replied:
saw the change in the following commit, igonre it

ramakrishnapallala commented on arch/xtensa/include/kernel_arch_data.h:
```diff
@@ -57,6 +57,10 @@ typedef struct _kernel_arch _kernel_arch_t;
 
 #endif /*! _ASMLANGUAGE && ! __ASSEMBLER__ */
 
+#ifdef CONFIG_USE_SWITCH
```
can we change the config name to CONFIG_USE_ARCH_SWITCH?

andrewboie replied:
why is that necessary?

ramakrishnapallala commented on kernel/include/ksched.h:
```diff
@@ -32,6 +32,13 @@ extern s32_t _ms_to_ticks(s32_t ms);
 
+/* find which one is the next thread to run */
+/* must be called with interrupts locked */
+static ALWAYS_INLINE struct k_thread *_get_next_ready_thread(void)
```
this change was removed earlier commit and added now again!!

ramakrishnapallala commented on arch/xtensa/core/xtensa-asm2.c:
```diff
@@ -9,6 +9,7 @@
 #include <kernel.h>
 #include <ksched.h>
 #include <kernel_structs.h>
+#include <nano_internal.h>
```
kernel_internal.h?

ramakrishnapallala commented on tests/kernel/mp/src/main.c:
```diff
@@ -0,0 +1,48 @@
+{
+	cpu_arg = 12345;
+
+	_arch_start_cpu(1, cpu1_stack, CPU1_STACK_SIZE, cpu1_fn, &cpu_arg);
```
does cpu1_fn() will be called from a new thread context?

ramakrishnapallala commented on kernel/include/kernel_structs.h:
```diff
@@ -89,6 +88,30 @@ struct _kernel {
+	 * the "_kernel.XXX" expressions and assembly offsets.
+	 */
+	union {
+		struct _cpu cpus[CONFIG_MP_NUM_CPUS];
```
strcut _cpu should be under #ifdef CONFIG_SMP right?

andyross replied:
No, this is the official API now.  When there is only one CPU, the array is of length one but otherwise shares the same offsets as the legacy fields in the union (which will eventually go away).  There's no particular benefit to having a separate API beyond making migration easier.

ramakrishnapallala commented on kernel/init.c:
```diff
@@ -302,11 +340,19 @@ static void prepare_multithreading(struct k_thread *dummy_thread)
+
+#if defined(CONFIG_SMP) && CONFIG_MP_NUM_CPUS > 3
+	init_idle_thread(_idle_thread3, _idle_stack3);
 #endif
```
all the #if should be under the CONFIG_MULTITHREADING

andyross replied:
CONFIG_SMP presumes CONFIG_MULTITHREADING.  It doesn't make sense otherwise.  Not sure if the kconfig makes that explicit or not.

ramakrishnapallala commented on kernel/smp.c:
```diff
@@ -48,3 +50,56 @@ void _smp_global_unlock(unsigned int key)
+{
+	atomic_t start_flag;
+
+	atomic_clear(&start_flag);
```
Just for my understanding, why we need atomic here?

andyross replied:
We don't on ESP-32 or x86.  In theory on some architecture we might if there are cache coherence rules..  Beyond that it's just form adherence.  Spinning on a raw int from a separate CPU just feels wrong.

ramakrishnapallala commented on kernel/include/ksched.h:
```diff
@@ -46,7 +53,11 @@ static inline int _is_idle_thread(void *entry_point)
 static inline int _is_idle_thread_ptr(k_tid_t thread)
 {
+#ifdef CONFIG_SMP
+	return thread->base.is_idle;
```
if we are checking idle thread ptr based on 'is_idle' then any thread which sets this filed becomes idle thread, do we want this ? idle thread is something which should be run at the system level and should be created early init right?

andyross replied:
Nothing sets this field at runtime, and threads do not "become idle".  An "idle thread" is a specific thread created at system startup.  There used to be one, so you would test it with "tid == _idle_thread".  Now there needs to be one per CPU so it needs a bit somewhere in the struct to tell us.

ramakrishnapallala commented on kernel/sched.c:
```diff
@@ -489,3 +497,44 @@ int _impl_k_is_preempt_thread(void)
+			struct k_thread *th = (struct k_thread *)node;
+
+			/* Skip threads that are already running elsewhere! */
+			if (th->base.active && th->base.cpu != mycpu) {
```
what does active flag indicate? does it mean thread has run at least once? also when are we assigning the cpu? is it during thread create?

andyross replied:
Documented in the struct as "Non-zero when actively running on a CPU".  This indicates that another CPU is actively running this thread.

And we don't assign threads to CPUs.  The "S" in SMP means that every CPU can run every process.  They select the next process from the scheduler in this code when they are about to context switch (i.e. someone has called k_yield(), or we are returning from handling an interrupt of a preemptible thread, etc...).

ramakrishnapallala commented on kernel/idle.c:
```diff
@@ -164,6 +164,20 @@ void idle(void *unused1, void *unused2, void *unused3)
+		while (1) {
+			k_busy_wait(100);
+			k_yield();
+		}
```
this means, scheduler code always run on main cpu? also as per the current logic other cpu's don't don't enter sleep states?

andyross replied:
Timeout code always runs on the main CPU, yes.  Right now it's based on counting ticks instead of finer-grained timeouts, so there's no easy way to translate that to a different CPU's interrupts, which happen at a different phase.   Scheduler code runs on every CPU, it's SMP :)

And yes: right now auxiliary CPUs can't enter a low power state because there is nothing to wake them up appropriately when a new thread enters the scheduler.  We have to choose either bad latency (i.e. have them wake up at their next timer tick interrupt, which can be millions of cycles late) or bad power behavior (spin on the scheduler looking for updates), and I picked the latter.

In the intermediate term we need an architecture where adding a schedulable thread to the run queue where there are idle CPUs will produce an IPI to wake up the others CPUs.

ramakrishnapallala replied:
got it. Thanks.

ramakrishnapallala commented on kernel/thread_abort.c:
```diff
@@ -37,13 +37,17 @@ void _impl_k_thread_abort(k_tid_t thread)
-		CODE_UNREACHABLE;
-	}
+	if (_is_in_isr()) {
+		irq_unlock(key);
```
we are just unlocking here and returning, where is reschedule is being called?

the earlier logic is now moved to ela=se part (non isr), so how this call will behave in case of ISR?

andyross replied:
The reschedule doesn't get called.  That's the whole point: in an ISR, you can't _Swap() on xtensa (and ARM too) because you then won't get to the interrupt exit code, which will then inspect the scheduler and return into whatever the next thread should be.  This is all explained in the commit message detailing that change.

ramakrishnapallala replied:
got it :), the reschedule will happen anyways after servicing the interrupt.

ramakrishnapallala commented:
@andy I think it's better to add the documentation as you have written detailed commit messages  :)

andyross commented:
Just series cleanup in this last revision, no code change.  Ram found an embarassing spot where I'd independently written both a patch and its precise revision (moving a function out of a header, then right back in) without realizing it.  Remove those two patches.

rgundi commented:
Andy, in the current form, this still breaks Intel S1000 compilation (due to missing _soc_inthandlers.h). Is it not possible to make the changes explicitly under the "XTENSA_ASM2" macro? Meaning, if this macro is undefined, the compilation should continue in the old style (where it doesn't look for _soc_inthandlers.h).

SebastianBoe reviewed (approved):
CMake-related changes are OK

andyross commented:
Yes, older platforms are absolutely supposed to be still supported as-is.  You're right that the xtensa-asm2.c file needs a header that doesn't exist though, which is a bug.  I can move the asm2 sources (which are supposed to build on their own) into a zephyr_sources_ifdef(), though I can't test out of tree stuff well.  It might be simpler to do this in the S1000 patch series.

andyross commented:
@rgundi : Might be simpler to just have you test.  This should resolve the missing header build failure:

``diff --git a/arch/xtensa/core/CMakeLists.txt b/arch/xtensa/core/CMakeLists.txt
index 9b2ae4e..672d897 100644
--- a/arch/xtensa/core/CMakeLists.txt
+++ b/arch/xtensa/core/CMakeLists.txt
@@ -3,6 +3,9 @@ zephyr_sources(
        cpu_idle.c
        fatal.c
        window_vectors.S
+       )
+
+zephyr_sources_ifdef(CONFIG_XTENSA_ASM2
        xtensa-asm2-util.S
        xtensa-asm2.c
        )

Commit:
xtensa/asm2: Don't needlessly build asm2 sources

Non-asm2 devices without a generated SoC interrupt file will see a
compile failure due to the missing header.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Nah, that's silly.  I can trivially validate that patch even if I can't build S1000.  Included.

rgundi commented:
The compilation passed that stage but failed later with the below message.
kernel_arch_func.h: Assembler messages:
kernel_arch_func.h:35: Error: unknown opcode or format name 'rsr.misc0'

So, I see the code "_arch_curr_cpu()" gets included even in case where XTENSA_ASM2 is not enabled. I tried excluding it but it threw other errors as it is referenced elsewhere. So, as discussed earlier, guess it's better to replace MISC0 with EXCSAVE2 which is present in all Xtensa SoCs

andyross commented:
@rgundi: Yeah, let's take this offline and deal with the stuff specific to out-of-tree patches separately.  Until I get a working toolchain here (and ideally hardware that matches) we're constantly going to be bumping into issues like this.

FWIW: it should absolutely be possible to disable compilation of that function and whatever else depends on it when ASM2 is unset.

Commit:
xtensa: Make _arch_curr_cpu() work outside SMP

The xtensa headers use this for simplicity when SMP is not enabled.
It should still build on older platforms that don't include the
asm2-style CPU pointer scheme.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Heh, same deal.  Said "let's do this later", then went in to see if I could synthesize the error locally, and it turned out to be trivial.  Patch included.

lpereira commented on arch/xtensa/soc/esp32/esp32-mp.c:
```diff
@@ -0,0 +1,192 @@
+
+#define _REG(base, off) (*(volatile u32_t *)((base) + (off)))
+
+#define RTC_CNTL_BASE             0x3ff48000
```
Please use the register macros from esp-idf. All ESP32 drivers are using them; I'd like to keep the consistency.

andyross replied:
Sorry, my reply must have been in the other PR.  The problem is that the existing API for these registers is sort of a mess, it doesn't express the register as an lvalue and wants to use a whole bunch of setter/getter macroness, and IIRC I had trouble making that work.  But I'll investigate again.

lpereira commented on arch/xtensa/soc/esp32/linker.ld:
```diff
@@ -30,6 +30,9 @@ PROVIDE ( esp32_rom_uart_attach = 0x40008fd0 );
 PROVIDE ( esp32_rom_intr_matrix_set = 0x4000681c );
 PROVIDE ( esp32_rom_gpio_matrix_in = 0x40009edc );
 PROVIDE ( esp32_rom_gpio_matrix_out = 0x40009f0c );
+PROVIDE ( esp32_rom_Cache_Flush = 0x40009a14 );
```
Nitpick: lower case names, please.

andyross replied:
That's the name in the Espressif header though, even though we don't use it from there.

andrewboie replied:
nothing like mixing up hump notation/CamelCase and underscore notation.
I say leave it to be consistent with the HAL.

lpereira commented on arch/xtensa/soc/esp32/esp32-mp.c:
```diff
@@ -0,0 +1,192 @@
+
+	appcpu_start();
+
+	while (!alive_flag) {
```
Are there any error conditions we should check while waiting for the CPU to be brought up? Do we need to panic on timeout?

andyross replied:
...maybe?  There are no errors reported by the existing API and the FreeRTOS integration didn't check IIRC.   I guess my feeling here would be "don't test for fatal errors, failure is OK if recovery is impossible".  A hung boot if the second CPU fails seems.... sort of OK to me?

lpereira commented on include/spinlock.h:
```diff
@@ -0,0 +1,69 @@
+# ifdef CONFIG_DEBUG
+	l->saved_key = k.key;
+# endif
+	while (!atomic_cas(&l->locked, 0, 1)) {
```
I'm not sure if it would be an issue in the systems Zephyr runs on (esp. on ESP32, where this makes sense now), but it's often a good thing to perform the locking primitive of a spinlock like this in order to reduce cache contention:

```C
while (!atomic_cas(&l->locked, 0, 1)) {
     while (l->locked != 0) {
     }
}
```

Note that `atomic_get()` is not used here on purpose: the generic implementation using GCC atomic builtins will enforce total ordering, which isn't needed because `atomic_cas()` will do that for us if the value ever changes.

andyross replied:
Even on x86, the resulting lock cmpxchg is safe from cache spamming.  The line goes into a S(hared) state in L1D the first time through the loop and the core hits L2/L3 only if it snoops someone else's change to that cache line.

But really long term if we care about performance what we'll do is to have each architecture code its own spinlock implementation with inlines instead of calling out to the generic atomic API.

lpereira commented on kernel/init.c:
```diff
@@ -373,6 +436,10 @@ FUNC_NORETURN void _Cstart(void)
+	smp_init();
+#endif
+
 	/* display boot banner */
```
Minor: It would be nice to have "SMP" somewhere in the boot banner.

andrewboie replied:
Good idea.

lpereira commented on include/kernel.h:
```diff
@@ -391,6 +391,17 @@ struct _thread_base {
+	u8_t is_idle;
+
+	/* Non-zero when actively running on a CPU */
+	u8_t active;
```
Do we need both `active` and `cpu`? Could `active` be a bitmap, where the nth-bit tells us this thread is running on the nth CPU? This would simplify the thread skipping logic in the scheduler a little bit.

andyross replied:
That's clever.  Yes, I think.  The "cpu" field might in principle be needed (by a fancier scheduler) to tell the kernel the CPU that a non-active thread was "last" scheduled on, but it isn't right now.

andrewboie commented on kernel/Kconfig:
```diff
@@ -549,6 +549,18 @@ config MAX_DOMAIN_PARTITIONS
+	primitive than the original __swap mechanism.  It is required
+	for an SMP-aware scheduler, or if the architecture does not
+	provide __swap.  In uniprocess situations where the
+	architecture provides both, _arch_switch incurs more somewhat
```
" _arch_switch incurs more somewhat overhead and may be slower"

Nitpick: awkward English.
Suggest: "_arch_switch incurs somewhat more overhead."

andyross replied:
I... don't know why that ifdef is there anymore.  At a guess: because it used to hold an inline function that relied on other stuff under USE_SWTICH and is now just a vestigial function declaration.

Will fix that and the phrasing.

andrewboie commented on kernel/include/kernel_internal.h:
```diff
@@ -51,30 +52,61 @@ extern void _setup_new_thread(struct k_thread *new_thread,
+	old_thread = _kernel.current;
 
-#ifdef CONFIG_STACK_SENTINEL
 	_check_stack_sentinel();
```
Me personally, I always preferred explicit ifdefs rather than defining various functions to empty macros, as I think it makes it clearer that this call site doesn't do anything.

I used to lose arguments with Dirk all the time about this.

Carry on, I'm just whining.

andrewboie commented on drivers/timer/xtensa_sys_timer.c:
```diff
@@ -465,6 +479,18 @@ void _xt_tick_divisor_init(void)
+	 * ticks by calling the handler function multiple times if the
+	 * delta was more than one _xt_tick_divisor.
+	 */
+	SET_TIMER_FIRE_TIME(GET_TIMER_CURRENT_TIME() + _xt_tick_divisor);
```
I had a fever dream that one day this driver was fixed up to use IRQ_CONNECT() and all the custom asm stuff for it was nuked from orbit. I should file a bug on that.

andrewboie commented on arch/xtensa/core/xtensa_intgen.py:
```diff
@@ -0,0 +1,129 @@
+
+# Pass an Xtensa core-isa.h file on stdin or the command line, emits a
+# C file on output containing optimized interrupt dispatch routines.
+
```
Suggest filing a user story to have someone with CMake chops make this part of the build.
I have no idea on how to do it either.

andrewboie commented on arch/xtensa/core/xtensa-asm2.c:
```diff
@@ -27,23 +28,23 @@ void *xtensa_init_stack(int *stack_top,
 	memset(bsa, 0, bsasz);
 
-	bsa[BSA_PC_OFF/4] = entry;
+	bsa[BSA_PC_OFF/4] = _thread_entry;
```
Oh jeez. I missed this in the initial implementation.
I'm a little amazed our test cases pass, one of the essential things _thread_entry does is call k_thread_abort() after main() returns.
Good stuff.

andrewboie commented on arch/xtensa/core/xtensa-asm2-util.S:
```diff
@@ -265,6 +265,30 @@ DEF_EXCINT 6, _handle_excint, xtensa_int6_c
+ * from the C handler instead, but that would add a LARGE overhead to
+ * some alloca() calls (those whent he caller has been spilled) just
+ * to save these five cycles during other exceptions and L1
+ * interrupts.  Maybe revisit at some point, with better benchmarking.
```
In commit message you mention some possible future optimizations. I suggest filing a low-priority user story about it so we don't forget. We could give it to a junior engineer or something.

andrewboie commented on kernel/smp.c:
```diff
@@ -48,3 +50,56 @@ void _smp_global_unlock(unsigned int key)
 	k_spin_unlock(&global_spinlock, sk);
 }
+
+extern k_thread_stack_t _interrupt_stack1[];
```
I'd really like to see these interrupt stacks and idle thread objects, etc. expressed as an array sized by some Kconfig, rather than hard-coded in this way.
We can punt to a future enhancement if needed.

andrewboie commented on include/spinlock.h:
```diff
@@ -0,0 +1,69 @@
+	 * setting a zero and (because we hold the lock) know the existing
+	 * state won't change due to a race.  But some architectures need
+	 * a memory barrier when used like this, and we don't have a
+	 * Zephyr framework for that.
```
We have a memory barrier macro which should enforce ordering of memory access:

```
include/toolchain/gcc.h
331-#endif
332-
333:#define compiler_barrier() do { \
334-	__asm__ __volatile__ ("" ::: "memory"); \
335-} while ((0))
```

Or do you need something more like an instruction barrier?

andyross replied:
That comment was sort of obtuse.  I mean a hardware barrier.  Architectures with relaxed cache coherence rules (none of ours, I'm told that older NUMA systems in the 80's did this) wouldn't expose a C assignment (i.e." locked = 0;") to the cache in such a way that other CPUs would see the lock released, so another could come along and get an inconsistent result.  Using the atomic API guarantees that we're using a correct primitive as long as the architecture does it right.

andrewboie commented on include/spinlock.h:
```diff
@@ -0,0 +1,69 @@
+	 * implementation.  The "irq_lock()" API in SMP context is
+	 * actually a wrapper for a global spinlock!
+	 */
+	k.key = _arch_irq_lock();
```
Just for my own understanding: this locks interrupts just on the current CPU or all of them in this context?
If this is global I might be being thick but I am unsure how this works.

andyross replied:
Only on the current CPU, because it's in a critical section while the spinlock is held.  The other CPUs can do their own thing until/unless they want to contend for the same lock.

andrewboie commented on arch/xtensa/soc/esp32/esp32-mp.c:
```diff
@@ -66,6 +68,13 @@ static void appcpu_entry2(void)
 	__asm__ volatile("rsync");
 
+	/* Set up the CPU pointer.  Really this should be xtensa arch
+	 * code, not in the ESP-32 layer
```
Was there some difficulty implementing it there?

andrewboie commented on kernel/idle.c:
```diff
@@ -164,6 +164,20 @@ void idle(void *unused1, void *unused2, void *unused3)
+	/* Simplified idle for non-default SMP CPUs pending driver
+	 * support.  The busy waiting is needed to prevent lock
+	 * contention.  Long term we need to wake up idle CPUs with an
+	 * IPI.
```
suggest filing a user story about this and other future improvements, possibly for others to work on as the team has bandwidth.

andrewboie commented on include/misc/__assert.h:
```diff
@@ -67,6 +67,11 @@
 #endif
 
+#ifdef CONFIG_FORCE_NO_ASSERT
+#undef __ASSERT_ON
```
Nasty. I can only imagine how much time was sunk tracking this down.
Suggest an upstream report to gcc team if you haven't already.

andrewboie commented on samples/mpu/mpu_stack_guard_test/sample.yaml:
```diff
@@ -5,7 +5,7 @@ tests:
     harness: console
     # We do not handle segfaults here
-    arch_exclude: posix riscv32 nios2
+    arch_exclude: posix riscv32 nios2 xtensa
```
This change is fine. I need to fix this filter so that it instead filters on whether the CPU has stack protection features.

tyrantlizard77 reviewed (approved):
I think we can ship this. Just some questions/suggestions.
